### ISSUE TITLE
failed to provision volume with StorageClass "nfs-storage-provisioner": unable to create directory to provision new pv: permission denied

### ISSUE THREAD

**Issue by etoews** [RedHatOfficial/ocp4-helpernode#41]
Thanks a lot for providing this helpernode code! It's given us a huge head start on deploying OpenShift 4.

I'm having a problem that's very similar to #35. I also added a separate disk to my VM and mounted it at `/export`. The exception being that fix for that issue didn't work for me.

```bash
$ cat /etc/exports
/export	*(rw,sync,root_squash)

$ showmount -e localhost
Export list for localhost:
/export *
```

I deployed the nfs-provisioner

```bash
$ helpernodecheck nfs-setup
Now using project "nfs-provisioner" on server "https://api.sandbox.edpay.nz:6443".

You can add applications to this project with the 'new-app' command. For example, try:

    oc new-app django-psql-example

to build a new example application in Python. Or use kubectl to deploy a simple Kubernetes application:

    kubectl create deployment hello-node --image=gcr.io/hello-minikube-zero-install/hello-node

Already on project "nfs-provisioner" on server "https://api.sandbox.edpay.nz:6443".
serviceaccount/nfs-client-provisioner created
clusterrole.rbac.authorization.k8s.io/nfs-client-provisioner-runner created
clusterrolebinding.rbac.authorization.k8s.io/run-nfs-client-provisioner created
role.rbac.authorization.k8s.io/leader-locking-nfs-client-provisioner created
rolebinding.rbac.authorization.k8s.io/leader-locking-nfs-client-provisioner created
securitycontextconstraints.security.openshift.io/hostmount-anyuid added to: ["system:serviceaccount:nfs-provisioner:nfs-client-provisioner"]
deployment.extensions/nfs-client-provisioner created
storageclass.storage.k8s.io/nfs-storage-provisioner created
storageclass.storage.k8s.io/nfs-storage-provisioner annotated
Now using project "default" on server "https://api.sandbox.edpay.nz:6443".

Deployment started; you should monitor it with "oc get pods -n nfs-provisioner"
```

Everything is fine with that.

```bash
$ oc get pods -n nfs-provisioner
NAME                                      READY   STATUS    RESTARTS   AGE
nfs-client-provisioner-7c84684645-grwj7   1/1     Running   0          76s
```

I create the registry PVC but get an error in the events.

```bash
$ oc create -f /usr/local/src/registry-pvc.yaml -n openshift-image-registry
persistentvolumeclaim/registry-pvc created

$ oc describe pvc registry-pvc -n openshift-image-registry
Name:          registry-pvc
Namespace:     openshift-image-registry
StorageClass:  nfs-storage-provisioner
Status:        Pending
Volume:
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: nfs-storage
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:
Access Modes:
VolumeMode:    Filesystem
Mounted By:    <none>
Events:
  Type     Reason                Age                   From                                                                                      Message
  ----     ------                ----                  ----                                                                                      -------
  Normal   Provisioning          39s (x4 over 2m24s)   nfs-storage_nfs-client-provisioner-7c84684645-grwj7_bdf62238-8371-11ea-b836-0a580a800207  External provisioner is provisioning volume for claim "openshift-image-registry/registry-pvc"
  Warning  ProvisioningFailed    39s (x4 over 2m24s)   nfs-storage_nfs-client-provisioner-7c84684645-grwj7_bdf62238-8371-11ea-b836-0a580a800207  failed to provision volume with StorageClass "nfs-storage-provisioner": unable to create directory to provision new pv: mkdir /persistentvolumes/openshift-image-registry-registry-pvc-pvc-52fb821f-61b9-48b1-b23b-8099223c5840: permission denied
  Normal   ExternalProvisioning  13s (x11 over 2m25s)  persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "nfs-storage" or manually created by system administrator
```

I'm deploying OpenShift 4.3.8 and this VM happens to be Oracle Linux 7 so the user is nfsnobody.

```
$ oc version
Client Version: 4.3.8
Server Version: 4.3.8
Kubernetes Version: v1.16.2

$ cat /etc/os-release
NAME="Oracle Linux Server"
VERSION="7.6"
ID="ol"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.6"
PRETTY_NAME="Oracle Linux Server 7.6"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:oracle:linux:7:6:server"
HOME_URL="https://linux.oracle.com/"
BUG_REPORT_URL="https://bugzilla.oracle.com/"

ORACLE_BUGZILLA_PRODUCT="Oracle Linux 7"
ORACLE_BUGZILLA_PRODUCT_VERSION=7.6
ORACLE_SUPPORT_PRODUCT="Oracle Linux"
ORACLE_SUPPORT_PRODUCT_VERSION=7.6

$ ll -d /export
drwxrwxrwx. 3 nfsnobody nfsnobody 4096 Mar  2 17:10 /export

$ sudo restorecon -Rv /export
restorecon reset /export context system_u:object_r:unlabeled_t:s0->system_u:object_r:usr_t:s0
restorecon reset /export/lost+found context system_u:object_r:unlabeled_t:s0->system_u:object_r:usr_t:s0

$ ls -ldZ /export/
drwxrwxrwx. nfsnobody nfsnobody system_u:object_r:usr_t:s0       /export/
```

I can hop into the nfs-client-provisioner pod and poke around a bit.

```bash
$ oc project nfs-provisioner
Already on project "nfs-provisioner" on server "https://api.sandbox.edpay.nz:6443".

$ oc rsh pod/nfs-client-provisioner-7c84684645-grwj7

/ # whoami
root

/ # ls -ld /persistentvolumes/
drwxrwxrwx    3 nobody   nobody        4096 Mar  2 04:10 /persistentvolumes/

/ # ls /persistentvolumes/
lost+found

/ # touch /persistentvolumes/hello.txt
touch: /persistentvolumes/hello.txt: Permission denied

/ # mkdir /persistentvolumes/test
mkdir: can't create directory '/persistentvolumes/test': Permission denied
```

So `/export` is owned by nfsnobody as it should be in OL 7 but `/persistentvolumes` is owned by nobody in the nfs-client-provisioner pod.

1. Could that be the source of the "permission denied" issue?
1. If so, any suggestions on how to troubleshoot or fix it?
1. If not, any ideas on what else it could be?

I tried changing the owner of `/export` to nobody and poked around in the pod again but `/persistentvolumes` is owned by 99 in the nfs-client-provisioner pod.

```bash
$ sudo chown nfsnobody:nfsnobody -R /export

$ oc rsh pod/nfs-client-provisioner-7c84684645-grwj7

/ # ls -al
...
drwxrwxrwx    3 99       99            4096 Mar  2 04:10 persistentvolumes

/ # ls /persistentvolumes/
ls: can't open '/persistentvolumes/': Permission denied
```

I reverted back to `/export` being owned by nfsnobody.

**Comment by christianh814** [comment]
@etoews Haven't tested this on OLS 7, but it should "just work"

Do you see anything of value running `journalctl -f` while trying to write the file? Also look into `/var/log/audit/audit.log` for things.

If you have `setroubleshoot-server` installed you can run `sealert -a /var/log/audit/audit.log`

**Comment by etoews** [comment]
Seems to me like it should "just work" too. I feel like I maybe missed something but haven't been able to pinpoint it yet.

Thanks for trying to reproduce it and the debugging suggestions.

Strangely there was no additional output at all in `journalctl -f` or `tail -f /var/log/audit/audit.log` when I ran `oc create -f /usr/local/src/registry-pvc.yaml -n openshift-image-registry`

Nothing amiss according to `sealert` either.

```
$ sudo sealert -a /var/log/audit/audit.log
100% done
found 0 alerts in /var/log/audit/audit.log
```

I set the security context for `/export` to match yours but still got the same "Permission denied." result.

```
$ sudo chcon -u unconfined_u /export

$ ll -dZ /export
drwxrwxrwx. nfsnobody nfsnobody unconfined_u:object_r:usr_t:s0   /export
```

I'm stumped at the moment. This is just a proof of concept cluster so I'm going to workaround this for the time being with emptyDir and revisit this later.

```
oc patch configs.imageregistry.operator.openshift.io cluster --type merge --patch '{"spec":{"storage":{"emptyDir":{}}}}'
```

I'll comment again if I ever come across the answer but I'm willing to bet if I rebuild from scratch it'll just work.
